src/gg.regression.unity.bots/Runtime/Scripts/ClientDashboard/RGTcpManager.cs

```diff
@@ -0,0 +1,245 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net.Sockets;
+using JetBrains.Annotations;
+using RegressionGames.RemoteOrchestration.Models;
+using RegressionGames.RemoteOrchestration.Types;
+using RegressionGames.StateRecorder;
+using RegressionGames.StateRecorder.BotSegments;
+using RegressionGames.StateRecorder.BotSegments.Models;
+using UnityEditor;
+using UnityEngine;
+
+namespace RegressionGames.ClientDashboard
+{
+    /// <summary>
+    /// Starts and stops RGTcpServer, handles and sends messages to and from connected clients
+    /// </summary>
+    [ExecuteAlways]
+    public class RGTcpManager : MonoBehaviour
+    {
+        // the last active sequence that was sent to the client
+        private BotSegmentsPlaybackController m_botSegmentsPlaybackController;
+        private BotSequence m_startPlayingSequence = null;
+        private ActiveSequence m_activeSequence = null;
+        private ReplayToolbarManager m_replayToolbarManager;
+        private List<AvailableBotSequence> m_availableBotSequences = new ();
+
+        /// <summary>
+        /// Create a menu item to open the client dashboard, which will connect to this server
+        /// </summary>
+        [MenuItem("Regression Games/Open Dashboard")]
+        public static void OpenRGDashboard()
+        {
+            // Configure the process
+            var startInfo = new ProcessStartInfo
+            {
+                FileName = Path.GetFullPath("Packages/gg.regression.unity.bots/Runtime/Resources/RegressionGames.exe"),
+                // TODO: Arguments,
+                CreateNoWindow = false
+            };
+            Process.Start(startInfo);
+        }
+        
+        public void Start()
+        {
+            EditorApplication.playModeStateChanged -= OnPlayModeStateChanged;
+            EditorApplication.playModeStateChanged += OnPlayModeStateChanged;
+
+            if (!Application.isPlaying)
+            {
+                // I don't know why Unity insists on instantiating this object to default fields
+                // but we really do want null...
+                m_activeSequence = null;
+            }
+            
+            RGTcpServer.Start();
+            RGTcpServer.OnClientHandshake -= OnClientHandshake;
+            RGTcpServer.OnClientHandshake += OnClientHandshake;
+            RGTcpServer.ProcessClientMessage -= ProcessClientMessage;
+            RGTcpServer.ProcessClientMessage += ProcessClientMessage;
+            StartCoroutine(RGSequenceManager.ResolveSequenceFiles(ProcessResolvedSequences));
+        }
+        
+        /// <summary>
+        /// Assign or reset variables when the editor transitions to or from play mode
+        /// </summary>
+        private void OnPlayModeStateChanged(PlayModeStateChange state)
```

Comment:
I don't know this for certain, but I would bet that the 'i can't keep static things across a reload' is less to do with play vs edit mode and more to do with the code compilation and the context reset that happens because of it.

I would suggest we experiment with ...

triggers around `CompilationPipeline.compilationStarted` and `CompilationPipeline.compilationFinished`


I realize that is in the UnityEditor namespace and thus limited to functionality in the editor.


We may be able to make our lives easier here by separating out the management of the TCP server to editor specific and build specific code.

Make the build specific code do nothing if it is in the editor that we're playing, and put the editor specific code in the Editor package itself.


I am also curious if this alone would help us solve the instance problem if we were to make the tcp server variables instance variables instead of static, and manage a static on the editor class itself that represents the active TCPServer object.

Reply:
After discussing this, ... this PR will just handle adding the # checks so that production builds still work outside the editor on merge

Reply:
Okay I got a working build of BossRoom with my changes. The dashboard (client no.3) doesn't seem to be able to play sequences or receive the active sequence in a build, but there isn't currently a way for the user to open the dashboard in a build anyway... something we'll definitely need to address but later.

Also hooked into the `DidReloadScripts` editor event to make sure the client always receives necessary data on editor start.

Reply:
suggestions for ways to open in a build you've likely already considered
- in game hotkey
- manually launch it from outside the game